### PR TITLE
fix(bundler): handle imports from inlined modules in entry functions

### PR DESCRIPTION
**Pull request by tinovyatkin**
## Summary

This PR fixes an issue where the bundler failed to properly handle imports from inlined modules when those imports occurred inside functions in the entry module. This was causing the `xfail_mixed_import_patterns` test to fail with `ModuleNotFoundError`.

## Changes

- Added detection for imports from inlined modules in the `rewrite_import_from` function
- Implemented new `handle_imports_from_inlined_module` method that creates direct assignments to renamed symbols
- Fixed the bundler to properly resolve symbol renames for inlined modules

## Test Results

- ✅ All tests passing (`cargo test --workspace`)
- ✅ No clippy warnings (`cargo clippy --workspace --all-targets`) 
- ✅ `xfail_mixed_import_patterns` test now passes correctly

The fix handles complex scenarios with circular dependencies and mixed import patterns at both module and function levels.

🤖 Generated with [Claude Code](https://claude.ai/code)

___

## **CodeAnt-AI Description**
- Fixed the static bundler to correctly handle imports from inlined modules, especially when such imports occur inside functions of the entry module, by detecting inlined modules and generating direct symbol assignments.
- Added a new method to resolve symbol renames for inlined modules, ensuring that imports reference the correct renamed symbols and preventing runtime `ModuleNotFoundError`.
- Improved debug logging to clarify module import handling paths.
- Updated `pyproject.toml` to reorganize dependency groups, separating fixture dependencies and refining development dependencies, and enhanced the `uv` tool configuration for better environment and prerelease handling.

This PR resolves a critical bug in the bundler's handling of mixed import patterns involving inlined modules, particularly in complex scenarios with circular dependencies. It also improves project dependency management and development tooling configuration.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>code_generator.rs</strong><dd><code>Fix and enhance handling of imports from inlined modules in bundler</code></dd></summary>
<hr>

crates/cribo/src/code_generator.rs
<li>Added detection and handling for imports from inlined modules within <br>the <code>rewrite_import_from</code> function.<br> <li> Introduced a new method <code>handle_imports_from_inlined_module</code> to generate <br>direct assignments for imported symbols from inlined modules.<br> <li> Ensured proper resolution of symbol renames for inlined modules, <br>avoiding reliance on <code>sys.modules</code>.<br> <li> Improved debug logging for module import handling.<br>

</details>

  </td>
  <td><a href="https://github.com/ophidiarium/cribo/pull/140/files#diff-c161e86d0f1bfc649f3989ca702cb4cc7ef92f0488634273d367b2d00fa86e3d">+60/-1</a>&nbsp; &nbsp; </td>
</tr>
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Update dependency groups and uv tool configuration for development and <br>fixtures</code></dd></summary>
<hr>

pyproject.toml
<li>Updated dependency groups: moved <code>vale</code> to <code>dev</code>, created a new <code>fixtures</code> <br>group for fixture testing dependencies.<br> <li> Adjusted <code>[tool.uv]</code> settings: changed <code>default-groups</code> to include <code>dev</code> and <br><code>fixtures</code>, added <code>prerelease</code> and <code>required-environments</code> configuration.<br>

</details>

  </td>
  <td><a href="https://github.com/ophidiarium/cribo/pull/140/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711">+11/-3</a>&nbsp; &nbsp; </td>
</tr>
</table></td></tr></tr></tbody></table><details>
<summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

- **New Features**
  - Introduced semantic analysis of Python import usage, enabling precise tracking of runtime vs. deferred import contexts, type-checking-only dependencies, and import movability.
  - Added a transformer to automatically remove no-op statements (e.g., self-assignments, identity operations, redundant literals, and unnecessary pass statements) from Python code.
  - Enhanced module resolution and import classification to follow Python's rules, prioritize entry directories, and support namespace packages.
  - Improved dependency graph to track and analyze type-checking-only dependencies.

- **Bug Fixes**
  - Improved detection and handling of circular dependencies and error-handling import contexts.
  - Enhanced side-effect detection for function/class definitions and imports.

- **Documentation**
  - Major updates to technical documentation, including detailed project overview, architecture, usage examples, and workflow instructions.
  - Added implementation plan outlining phased development for semantic import analysis.

- **Configuration**
  - Added and updated configuration files for formatting, linting, testing, and CI (e.g., `.shellcheckrc`, `.markdownlint-cli2.yaml`, `.clippy.toml`, `.github/workflows/ci.yml`).
  - Updated VS Code settings and recommended extensions for improved developer experience.

- **Refactor**
  - Reorganized and streamlined import statements, logging, and configuration across multiple files.
  - Simplified and improved code structure for module resolution, dependency tracking, and visitor patterns.
  - Refined import rewriting and bundling logic to leverage semantic import information.

- **Tests**
  - Added comprehensive test fixtures and cases to verify semantic import analysis, no-op removal, and edge-case handling in multi-module Python projects.

- **Chores**
  - Updated CI workflows to use Nextest for testing and coverage.
  - Reformatted and reordered configuration files for clarity and consistency.

### CODE REVIEW

**Comment by tinovyatkin** [comment]
@coderabbit full review

File: crates/cribo/src/orchestrator.rs [hunk 1]
```diff
@@ -1080,24 +1225,57 @@ impl BundleOrchestrator {
...
+        // Legacy method - assumes imports are not in error handlers
+        // This is used by other parts of the code that don't track context
+        match self.process_import_for_discovery_with_context(import, false, params) {
+            Ok(()) => {}
+            Err(e) => {
+                // For backward compatibility, just warn instead of failing
+                warn!("Failed to process import: {e}");
             }
         }
     }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs:1282]: ✅ Addressed in b2c45b2. Thanks! 

Removed all the unused helper methods:
-  (replaced by )
- 
- 
- 
- 
- 

These were indeed remnants from refactoring and have been cleaned up.

[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs:1282]: ✅ Addressed in b2c45b2. Thanks\! 

Removed all the unused helper methods:
- process_import_for_discovery (replaced by process_import_for_discovery_with_context)
- build_cycle_error_message
- append_unresolvable_cycles_to_error
- append_resolvable_cycles_to_error
- append_cycle_resolution_suggestions
- log_resolvable_cycles

These were indeed remnants from refactoring and have been cleaned up.

File: crates/cribo/src/resolver.rs [hunk 2]
```diff
@@ -294,165 +340,201 @@
...
 
-        let site_packages = path.join("site-packages");
-        if site_packages.exists() && site_packages.is_dir() {
-            Some(site_packages)
-        } else {
-            None
-        }
+        // Default to third-party if we can't determine otherwise
+        ImportType::ThirdParty
     }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/resolver.rs]: ✅ Addressed in c8b6a09. Thanks\!

The `classify_import` method now takes `&mut self` and caches all classification results, avoiding repeated filesystem operations for the same module.

File: crates/cribo/src/visitors/side_effect_detector.rs [hunk 3]
```diff
@@ -143,13 +143,44 @@
...
         match stmt {
-            // These statements are pure definitions, no side effects
-            Stmt::FunctionDef(_) | Stmt::ClassDef(_) => {
-                // Don't recurse into function/class bodies
+            // Function definitions are pure, no side effects
+            Stmt::FunctionDef(_) => {
+                // Don't recurse into function bodies
                 // Their execution is deferred
                 return; // Important: don't call walk_stmt
             }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/visitors/side_effect_detector.rs]: ✅ Addressed in a079306. Thanks!

The side effect detector now properly checks:
- Decorator expressions
- Default argument values  
- Return and parameter annotations

All of these are executed at import time and can have side effects, as you correctly pointed out.

I also added special handling to exclude `typing` imports and properly handle type annotations to avoid false positives (e.g., `List[int]` in annotations is not a side effect).

File: crates/cribo/src/orchestrator.rs [hunk 4]
```diff
@@ -1421,12 +1421,22 @@ impl BundleOrchestrator {
...
-                    .analyze_movable_imports(params.graph, &analysis.resolvable_cycles);
+                // Prepare module ASTs for semantic analysis
+                let module_ast_pairs: Vec<(String, ModModule)> = module_asts
+                    .iter()
+                    .map(|(name, ast, _, _)| (name.clone(), ast.clone()))
+                    .collect();
+
+                // Analyze movable imports using semantic analysis
+                let movable_imports = import_rewriter.analyze_movable_imports_semantic(
+                    params.graph,
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/orchestrator.rs]: ✅ Addressed in d5d7073. Thanks\!

Changed `analyze_movable_imports_semantic` to accept `&[(String, &ModModule)]` instead of `&[(String, ModModule)]`, avoiding the unnecessary deep cloning of ASTs.

This reduces both memory usage and CPU overhead since the function only needs read-only access to the ASTs.

File: crates/cribo/src/import_rewriter.rs [hunk 5]
```diff
@@ -100,6 +103,128 @@ impl ImportRewriter {
...
+                }
+            }
+        }
+
+        debug!(
+            "Found {} movable imports using semantic analysis",
+            movable_imports.len()
+        );
+        Ok(movable_imports)
+    }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/import_rewriter.rs]: ✅ Addressed in 56f49a9. Thanks\!

Added caching for import discovery results using `FxHashMap<ModuleId, Vec<DiscoveredImport>>`. When a module appears in multiple circular dependency cycles, we now reuse the cached analysis results instead of re-running the semantic analysis.

This optimization reduces the complexity from O(N×C) to O(N), where N is the number of modules and C is the number of cycles, significantly improving performance for large dependency graphs with overlapping cycles.

File: crates/cribo/src/import_rewriter.rs [hunk 6]
```diff
@@ -100,6 +103,128 @@
...
+                            level: import_info.level,
+                        }
+                    };
+
+                    movable.push(MovableImport {
+                        import_stmt,
+                        target_functions,
+                        source_module: module_name.to_string(),
+                        line_number: 0, // TODO: Extract from import_info.import_location
+                    });
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/import_rewriter.rs]: ✅ Addressed in 7156c34. 

- Import aliases are now properly preserved by checking the structure of DiscoveredImport
- Line numbers now use the actual byte offset from TextRange.start() instead of hardcoding to 0
- The code correctly distinguishes between regular imports (e.g., 'import foo as bar') and from imports

Note: While byte offsets aren't actual line numbers, they provide a usable placeholder until we integrate proper line number computation from source text.

Thanks\!

File: crates/cribo/src/semantic_analysis.rs [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+            .module_name
+            .split('.')
+            .next_back()
+            .unwrap_or(&import.module_name);
+        visitor.register_import(import_name, &import.module_name);
+
+        // Also register any imported names from 'from' imports
+        for name in &import.imported_names {
+            visitor.register_import(name, &import.module_name);
+        }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/semantic_analysis.rs]: ✅ Addressed in 0b91894.

Updated the semantic analysis module to properly track import aliases:
- Changed `ImportInfo.imported_names` from `Vec<String>` to `Vec<(String, Option<String>)>` to preserve alias information
- Modified the visitor registration loop to register both the original name and its alias (if present)

Now when code uses an aliased import like:
```python
from mod import func as z
z()   # ← now properly recognized
```

Both "func" and "z" are registered with the visitor, ensuring all usages are tracked correctly.

Thanks for the suggestion\!

File: crates/cribo/src/semantic_import_context.rs [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+                        usages: vec![],
+                        has_side_effects: import_from
+                            .module
+                            .as_ref()
+                            .map(|m| Self::has_side_effects(m.as_str()))
+                            .unwrap_or(false),
+                        level: import_from.level,
+                        import_location: import_from.range,
+                    });
+            }
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/semantic_import_context.rs]: This code has been refactored since this comment was made. The file `semantic_import_context.rs` was deleted in commit 4878e70 and its functionality was moved to `visitors/import_discovery.rs`.

In the current implementation, multiple imports from the same module are correctly handled as separate entries in the imports vector. Each `from` import statement creates its own `DiscoveredImport` entry with its specific imported names, so the issue of losing data from repeated imports no longer exists.

For example:
```python
from pkg import a
from pkg import b
```

These now create two separate `DiscoveredImport` entries, each preserving their respective imported names.

Thanks for the suggestion though - it would have been a valid concern for the old implementation!

File: crates/cribo/src/graph_builder.rs [hunk 9]
```diff
@@ -208,7 +210,8 @@ impl<'a> GraphBuilder<'a> {
                 level: import_from.level,
                 is_star,
             },
-            var_decls: imported_names.clone(), // FromImport declares the imported names as variables
+            var_decls: imported_names.clone(), /* FromImport declares the imported names as
+                                                * variables */
             read_vars: FxHashSet::default(),
```
[the remark this answers is not in the document]
Comment (tinovyatkin) [reply crates/cribo/src/graph_builder.rs]: ✅ Addressed in c198e1d.

Applied the suggested fix to prevent '*' from being incorrectly declared as a variable when processing 'from x import *' statements.

Star imports now correctly result in an empty var_decls set, which properly represents that we don't know what variables will be declared until runtime.

Thanks for catching this\!
